### PR TITLE
Default CAE challenge handling in BearerTokenAuthenticationPolicy

### LINKED ISSUE #41984: BearerTokenAuthenticationPolicy should support CAE token revocation challenges by default
This feature entails adding CAE support for all clients lacking a custom challenge handler i.e., everyone except Key Vault and Storage. 

Adding support involves adding logic to your BearerTokenAuthenticationPolicy such that it does the following:
 - Detects when a CAE challenge is issued (401 response with a WWW-Authenticate header)
 - Parses the WWW-Authenticate header (format [here](https://learn.microsoft.com/entra/identity-platform/app-resilience-continuous-access-evaluation?tabs=dotnet#the-code))
   - validate that the `error` value is "insufficient_claims"
   - capture the `claims` value and decode it from base64 encoding to a string
 - Pass the string value of the un-encoded `claims` to the `TokenCredential` via the `TokenRequestContext` or equivalent for your language via the `Claims` property
 - Ensure that any local token caching is bypassed in the policy when the claims are populated from a CAE challenge
 - Authorize the original request with the new token and send it through the pipeline again
 - Return any response to the caller (don't try to handle a second challenge)

Example PRs: 
https://github.com/Azure/azure-sdk-for-go/pull/23414
https://github.com/Azure/azure-sdk-for-net/pull/46277

### PR DESCRIPTION
**Pull request by billwert**
This adds handling CAE challenges by default in our policy.

Fixes #41984

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:21]: Mind adding documentation with a few examples on what this will be looking for in case this ever needs to be troubleshooted.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:235]:
```suggestion
            LOGGER.warning("Failed to decode the claims from the CAE challenge. Encoded claims: " + encodedClaims);
```

To make it easier to read

Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:235]: Is it possible for claims to contain sensitive information that we shouldn't log by default but require additional configuration to enable their logging?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:231]: Should the try/catch migrate to within the if block? The catch statement can only be hit if the if block is entered.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: Is there a CHANGELOG entry we should add discussing that CAE challenge is enabled by default? Also, is there a way to disable it by default in case a bug scenario is hit?

Comment (weidongxu-microsoft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: There appears to be a small difference from `ArmChallengeAuthenticationPolicy` used for ARM CAE.
https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-management/src/main/java/com/azure/core/management/http/policy/ArmChallengeAuthenticationPolicy.java#L59

Though I assume they are handling for the same CAE challange.

What would happen to `ArmChallengeAuthenticationPolicy` after this PR merged?
Should mgmt lib switch to use `BearerTokenAuthenticationPolicy` directly (and deprecate `ArmChallengeAuthenticationPolicy`)?

Comment (christothes) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: I'll let @billwert comment when he returns, but I believe we should be able to replace `ArmChallengeAuthenticationPolicy` with `BearerTokenAuthenticationPolicy`. It is possible that the older policy is based on a slightly different format of the challenge.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (chlowell) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:235]: No; claims aren't secret (note they're sent to unauthorized clients)

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (weidongxu-microsoft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: Yes, I think this be the plan too. New SDK generated would use `BearerTokenAuthenticationPolicy` directly, while `ArmChallengeAuthenticationPolicy` would still be in place (maybe updated) for backward-compatibility.

One thing we need to verify is that `ArmChallengeAuthenticationPolicy` would still work after the change to `BearerTokenAuthenticationPolicy` (its base class).

Comment (christothes) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: I believe that the older implementation of CAE that `ArmChallengeAuthenticationPolicy` targeted was never actually implemented. Does it perform any other function besides handling CAE challenges? If not, there may be no reason to keep it for backwards compat.

Comment (chlowell) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: 👍 ARM CAE never left preview so there's no need to keep the client code we wrote for it years ago

Comment (weidongxu-microsoft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: > I believe that the older implementation of CAE that `ArmChallengeAuthenticationPolicy` targeted was never actually implemented. Does it perform any other function besides handling CAE challenges? If not, there may be no reason to keep it for backwards compat.

We actually verified the behavior of CAE in `ArmChallengeAuthenticationPolicy` via end2end test on a prod tenant in 2021 on ARM Microsoft.Resources RP. So, customer can use that since 2021; though I don't know how many customer indeed depends on this preview CAE feature.

Comment (chlowell) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: ARM disabled their CAE implementation early this year. It had no users at that time

Comment (billwert) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:81]: Thanks for the discussion while I was gone!

It sounds like the conclusion here is that we should remove `ArmChallengeAuthenticationPolicy` and use what we are adding here. @weidongxu-microsoft, do you want to do that in a follow up once this is in?

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (chlowell) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:31]: This is necessary but not sufficient--a CAE challenge also has a "claims" parameter

File: sdk/core/azure-core/src/test/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicyTests.java
Comment (chlowell) [inline sdk/core/azure-core/src/test/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicyTests.java:36]: encodedClaims isn't used

Comment (g2vinay) [reply sdk/core/azure-core/src/test/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicyTests.java:37]: updated it to be used.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:109]: I wonder if `AuthorizationChallengeParser` could return a richer object. Right now, if this conditional is met the response challenge will be parsed multiple times. The operation should be fairly cheap, but this may be one of the more commonly called code paths. For now, I think this is fine and something we can iterate on later.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:13]: nit: `final`

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:14]: nit: private constructor

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:82]: Get the value after the `key` check, this might just waste time

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:56]: nit: Could check

```java
int schemeIndex = challenge.indexOf(challengeScheme);
if (schemeIndex == -1) {
    return null;
}

if (challenge.charAt(schemeIndex + challengeScheme.length() + 1) != " ") {
    return null;
}
```

to get around creating a new string instance when concatenating the `" "`.

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:25]: This is fine for now, but in the future we should look into a way to get multiple parameter values without having to do duplicative parsing of the parameters

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:21]: Would we ever have support for a different header? Such as `Proxy-Authenticate`?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java
Comment (alzimmermsft) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/policy/BearerTokenAuthenticationPolicy.java:235]: This isn't an API in Java 8

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java
Comment (g2vinay) [reply sdk/core/azure-core/src/main/java/com/azure/core/implementation/http/policy/AuthorizationChallengeParser.java:27]: No plans as of now.

**Review by alzimmermsft — APPROVED** [review]

**Review by g2vinay — APPROVED** [review]

**Pull request merged**
